### PR TITLE
S2259: Enhance SE RuleFactory

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #5862

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs [hunk 1]
```diff
@@ -327,9 +328,25 @@ public void Method()
 
         [TestMethod]
         public void Analyze_Severity_ExecutesWhenAll() =>
-            Verify(@"string s = null;   // Noncompliant    {{Message for SAll}}
-                                        // Noncompliant@-1 {{Message for SMain}}
-                    s.ToString();       // Noncompliant    {{'s' is null on at least one execution path.}} - rule S2259");
+            VerifyCode<TestSERunnerCS>(@"
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs:331]: I moved to S2222. S2259 just caused too much trouble all the time.

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 2]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
...
+            where TCheck : SymbolicRuleCheck, new()
+        {
+            public RuleFactory() : this(null) { }
+            public RuleFactory(Func<object> sonarFallbackFactory) : base(typeof(TCheck), () => new TCheck(), sonarFallbackFactory) { }
         }
 
-        private sealed class RuleFactory<TCheck> : RuleFactory
+        protected class RuleFactory<TCheck, TSonarFallback> : RuleFactory<TCheck>
             where TCheck : SymbolicRuleCheck, new()
+            where TSonarFallback : new()
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:148]: It is probably a good idea to constrain this type parameter to some "marker" interface (and ISymbolicExecutionAnalyzer should derive from it). Otherwise, the "Roslyn" rule can be passed here, which is an easy mistake to do (same class name!).

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs:331]: As [expected](https://github.com/SonarSource/sonar-dotnet/pull/5904#discussion_r929686955) 😄

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs [hunk 3]
```diff
@@ -327,9 +328,25 @@ public void Method()
...
+using System.Threading;
+namespace Monitor_Conditions
+{
+    class Program
+    {
+        private object obj = new object();
+        public void Method1(bool condition)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs:338]: This scaffolding should not be needed. The rule will trigger for local variable too. Using `Verify(` as before should do the trick - adding the `using` on line 422 will solve the namespace issue.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs [hunk 4]
```diff
@@ -112,11 +111,11 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            var enabledAnalyzers = SonarRules.Where(x => x.SupportedDiagnostics.Any(descriptor => IsEnabled(context, isTestProject, isScannerRun, descriptor))).ToArray();
-            if (!Configuration.ForceSonarCfg) // ToDo: Will be removed in RuleFactory PR
-            {
-                enabledAnalyzers = enabledAnalyzers.Where(x => !x.SupportedDiagnostics.Any(descriptor => AllRules.Keys.Any(dd => descriptor.Id == dd.Id))).ToArray();
-            }
+            var allSonarRules = SonarRules.Union(AllRules
+                                .Select(x => x.Value.CreateSonarFallback(Configuration))
+                                .WhereNotNull()
+                                .Cast<ISymbolicExecutionAnalyzer>()); // Only ISymbolicExecutionAnalyzer should be passed as TSonarFallback to CreateFactory. Have you passed a Roslyn rule instead?
+            var enabledAnalyzers = allSonarRules.Where(x => x.SupportedDiagnostics.Any(descriptor => IsEnabled(context, isTestProject, isScannerRun, descriptor))).ToList();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs:118]: Let's do it the other way around to deal with the formatting issue. Calcullate AllRules and union with sonar:
```
            var allSonarRules = AllRules
                .Select(x => x.Value.CreateSonarFallback(Configuration))
                .WhereNotNull()
                .Cast<ISymbolicExecutionAnalyzer>() // Only ISymbolicExecutionAnalyzer should be passed as TSonarFallback to CreateFactory. Have you passed a Roslyn rule instead?
                .Union(SonarRules);
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs:118]: After that, we can continue with the original `Where` to calculate `enabledAnalyzers` in one go. The original `.ToList()` will need to go to its own line.

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 5]
```diff
@@ -81,8 +86,8 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
             var checks = AllRules
                 .Where(x => IsEnabled(nodeContext, isTestProject, isScannerRun, x.Key))
                 .GroupBy(x => x.Value.Type)                             // Multiple DiagnosticDescriptors (S2583, S2589) can share the same check type
-                .Select(x => x.First().Value.CreateInstance(sonarContext, nodeContext))   // We need just one instance in that case
-                .Where(x => x.ShouldExecute())
+                .Select(x => x.First().Value.CreateInstance(Configuration, sonarContext, nodeContext))   // We need just one instance in that case
+                .Where(x => x?.ShouldExecute() == true)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:90]: To highlight the null-checking component of it?
```suggestion
                .Where(x => x?.ShouldExecute() is true)
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 6]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
 
         protected class RuleFactory
         {
-            public Type Type { get; }
             private readonly Func<SymbolicRuleCheck> createInstance;
+            private readonly Func<object> sonarFallbackFactory;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:112]: To align naming with the previous line. It does (almost) the same thing.
```suggestion
            private readonly Func<object> createSonarFallbackInstance;
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 7]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
...
+
+            public Type Type { get; }
 
-            protected RuleFactory(Type type, Func<SymbolicRuleCheck> createInstance)
+            protected RuleFactory(Type type, Func<SymbolicRuleCheck> createInstance, Func<object> sonarFallbackFactory)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:116]: Same in the argument name

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 8]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
...
             {
                 Type = type;
                 this.createInstance = createInstance;
+                this.sonarFallbackFactory = sonarFallbackFactory;
             }
 
-            public SymbolicRuleCheck CreateInstance(SonarAnalysisContext sonarContext, SyntaxNodeAnalysisContext nodeContext)
+            public SymbolicRuleCheck CreateInstance(IAnalyzerConfiguration configuration, SonarAnalysisContext sonarContext, SyntaxNodeAnalysisContext nodeContext)
             {
+                if (configuration.ForceSonarCfg && sonarFallbackFactory != null)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:125]:
```suggestion
                if (configuration.ForceSonarCfg && sonarFallbackFactory is not null)
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 9]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
...
+                    return null;
+                }
+
                 var ret = createInstance();
                 ret.Init(sonarContext, nodeContext);
                 return ret;
             }
+
+            public object CreateSonarFallback(IAnalyzerConfiguration configuration) =>
+                configuration.ForceSonarCfg ? sonarFallbackFactory?.Invoke() : null;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:136]: Same thing, just lowering the brain load to understand the logic. The null-conditional makes it a bit harder to follow the intention.
```suggestion
                configuration.ForceSonarCfg && sonarFallbackFactory is not null ? sonarFallbackFactory() : null;
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 10]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
...
+        }
+
+        protected class RuleFactory<TCheck> : RuleFactory
+            where TCheck : SymbolicRuleCheck, new()
+        {
+            public RuleFactory() : this(null) { }
+            public RuleFactory(Func<object> sonarFallbackFactory) : base(typeof(TCheck), () => new TCheck(), sonarFallbackFactory) { }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:143]: This one should be protected for the system to make sense, right? 
```suggestion
            protected RuleFactory(Func<object> sonarFallbackFactory) : base(typeof(TCheck), () => new TCheck(), sonarFallbackFactory) { }
```

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs [hunk 11]
```diff
@@ -103,27 +108,46 @@ private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysis
...
         }
 
-        private sealed class RuleFactory<TCheck> : RuleFactory
+        protected class RuleFactory<TCheck, TSonarFallback> : RuleFactory<TCheck>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:146]: This chaining makes the constructors actually hard to follow and forces the `RuleFactory<TCheck>` to be aware of the sonar fallback - and it shouldn't know about it.  Let's change the inheritance for both to inherit directly `RuleCheck`.

There's slight duplication on the argument calculation, but it makes it more streight-forward in terms of understandability.
```cs
        protected class RuleFactory<TCheck> : RuleFactory
            where TCheck : SymbolicRuleCheck, new()
        {
            public RuleFactory() : base(typeof(TCheck), () => new TCheck(), null) { }
        }

        protected class RuleFactory<TCheck, TSonarFallback> : RuleFactory
            where TCheck : SymbolicRuleCheck, new()
            where TSonarFallback : new()
        {
            public RuleFactory() : base(typeof(TCheck), () => new TCheck(), () => new TSonarFallback()) { }
        }
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Some polishing

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs (hunk 11 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs:146]: `protected` gives a compiler error (`private protected` too). I changed the class to `private sealed` and made the actors `internal`.

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs (hunk 10 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs]: protected gives a compiler error (private protected too) on the ctor. I changed the class to private sealed and made the ctors internal.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs]: The `protected` comment was actually outdated after the next one below, because the this pass-through constructor was removed.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs]: `protected` class and `public` constructor is the desired state

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs (hunk 11 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs]: I don't see this one in the code - the constructors themselves should be changed and one of them removed.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Just finishing the previous round

File: analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs (hunk 10 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/SymbolicExecutionRunnerBase.cs]: Just to make sure: `private sealed class` was the original state. I changed it to `protected sealed`now.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs [hunk 12]
```diff
@@ -112,11 +111,12 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            if (!Configuration.ForceSonarCfg) // ToDo: Will be removed in RuleFactory PR
-            {
-                enabledAnalyzers = enabledAnalyzers.Where(x => !x.SupportedDiagnostics.Any(descriptor => AllRules.Keys.Any(dd => descriptor.Id == dd.Id))).ToArray();
-            }
+            var enabledAnalyzers = AllRules.Select(x => x.Value.CreateSonarFallback(Configuration))
+                                           .WhereNotNull()
+                                           .Cast<ISymbolicExecutionAnalyzer>() // ISymbolicExecutionAnalyzer should be passed as TSonarFallback to CreateFactory. Have you passed a Roslyn rule instead?
+                                           .Union(SonarRules)
+                                           .Where(x => x.SupportedDiagnostics.Any(descriptor => IsEnabled(context, isTestProject, isScannerRun, descriptor)))
+                                           .ToList();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs:119]: This should be indented 4 spaces according to coding style. Can be fix in any other PR

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
